_posts/2019-06-25-building-a-mono-repo.markdown

```diff
@@ -0,0 +1,66 @@
+---
+layout: post
+title:  "Building Multiple Projects in a Monorepo"
+date:   2017-04-06 17:23:53 -0700
+categories: jekyll update
+---
+
+<div align="center">
+<blockquote class="twitter-tweet"><p lang="en" dir="ltr">How do people set up Jenkins for their monorepos (... without converting everything to bazel/pants/etc; just many projects in one repo, with different build requirements, interdependent)?</p>&mdash; Dmitriy Ryaboy (@squarecog) <a href="https://twitter.com/squarecog/status/1143678611406241792?ref_src=twsrc%5Etfw">June 26, 2019</a></blockquote> <script async src="https://platform.twitter.com/widgets.js" charset="utf-8"></script>
+</div>
+
+# What is in your repo, and what does it build?
+
+The first question to ask your self is what is the nature of the projects in your monorepo, and what does "build" mean for them. Do you:
+* Build the contents of Pull Requests and gate merging them on the build passing?
+* Automatically do... "things" happen when changes are merged to master? Or what ever the special branch name you picked for your unnecessarily complicated git workflow.
+* Generate artifacts out of band from your release process and later have a user promote them? This can be through images such as an AMI or a Docker, more traditional OS packages like RPM's and Debs (the 90's called btw), or language specific packages (jars, wheels, gems, etc).
+* Perform an automated deployment of some set of project(s) without a human in the loop? Hopefully after after some level of automated testing has been performed, but hey this your repo. You yolo you.
+* Any of the above, but in repsonse to a dependent project being "built" (for that project's definition of build)?
+* Or, God forbid, all of the above?
+
+I pinged Dmitriy on Twitter after his initial post and he confirmed that he needed all of the above. You have my condolences, Dmitriy.
+
+
+# Ideas Informed by Experience
+
+I won't endorse any particular approach because:
```

Comment:
could skip these 3 lines for brevity, and just jump to "i'll explain my 3 ways w/ pros and cons"